### ISSUE TITLE
Perform commits on a separate repository

### ISSUE THREAD

**Issue by cliffchapmanrbx** [contributor-assistant/github-action#12]
Hi all,

We're looking very closely at this project for deploying a CLA blocking system to our repos. I'm pretty excited about the idea of not having to run a bot somewhere on our infrastructure just to get it working.

It looks like the bot will currently attempt to commit back to the same repository in order to add new approvals. How difficult would it be to implement a remote repository search instead of a local repository search? We'd prefer to keep the list of contributors both secret in a hidden repository as well as centralized so that contributors don't need to re-authenticate repeatedly for separate repositories.

Given the choice between 'spend developer effort implementing remote repo support for this project' and 'run through the hoops of securely deploying another external-facing service that has ongoing maintenance concerns' I can probably get clearance to write this feature myself, if it's something y'all would accept.

**Comment by ibakshay** [comment]
hi @cliffchapmanrbx, thank you very much for reaching out to us.  
Yes, this bot attempts to commit back to the same repository using built-in [GitHub Actions Token](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token). It has fixed token permissions and we cannot use this token to commit to a different repository. 

For remote repo support, You can use your [Personal Access Token (PAT)](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) with `repo admin` scope for committing to a remote repository. This PAT can be stored in the [secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets) and can be read from inputs in the workflow file. 
<img width="927" alt="Screenshot 2020-04-01 at 09 13 51" src="https://user-images.githubusercontent.com/33329946/78109398-63678100-73f9-11ea-8dd5-f529cfb37af2.png">
Remote repo support feature is not implemented yet, however, I believe that the development effort will be minimum. 
>  I can probably get clearance to write this feature myself

Yes, it will be great If you can raise a pull request for supporting the remote repo support feature. 
https://github.com/cla-assistant/github-action/blob/f2951bfe08846ae8c62b12420390a6500bf2394e/src/octokit.ts#L3
Tip: Here, you can have a condition statement something like, If the PAT is present then use PAT instead of `GITHUB_TOKEN`.   
Feel free to reach out to us again If you have any other questions.

**Comment by cliffchapmanrbx** [comment]
Hi @ibakshay, thanks for the writeup. It certainly helps point me in the right direction.

It looks like the tool is [just using Octokit](https://github.com/cla-assistant/github-action/blob/f2951bfe08846ae8c62b12420390a6500bf2394e/src/checkcla.ts#L71) to pull file contents, which just uses the standard GitHub API. I've used OctoKit (via .NET) previously so this should be straightforward.

One thing I'd like to look at in addition to using a PAT is using a GitHub App. Our Enterprise subscription to GitHub is pay-per-user, so if we need to create a service account for this there is a recurring cost to have that account. I'll see what I can do to come up with an implementation that can use either PATs or GitHub App integrations. I just want to make sure you have no objections to that approach.

I'll get started on that PR in the meantime.

**Comment by ibakshay** [comment]
@cliffchapmanrbx,  okay!! sounds great to me 💯. Creating a GitHub App on `github.com` is free and so you don't need to worry about any recurring costs. 
  Looking forward to your Pull Request 😄.

**Comment by cliffchapmanrbx** [comment]
Okay, I have a confession.

I started working on this feature and wanted to unit test it, and started having to write more code and more code to make it work. I'm not a JavaScript or TypeScript developer (I mostly work in .NET) so I had to learn a lot along the way to make it work.

Eventually I just started rewriting the code. The original code's processes worked great but I was having trouble understanding how the code actually worked. The result of my work is here: https://github.com/Roblox/cla-signature-bot

At this point I can't really open a pull request to upstream my changes, I rewrote pretty much all of the code in the process of unit testing everything. I could open a pull request but I'm sure you wouldn't appreciate it.

I've specifically called out your work @ibakshay as the foundation for what I ended up writing. If there's anything I can do to help contribute my work back into this project please let me know, and of course in the interest of the Apache license you are more than welcome to incorporate the improvements I've made. I can work on getting your CLA signed for my company.

**Comment by ibakshay** [comment]
Hi @cliffchapmanrbx, thank you very much for taking the time to inform me about the progress. I really appreciate it. I am very happy that you are able to implement this functionality and you also added unit tests. I  just had a look at your project. 

There is actually an open issue for implementing [unit tests](https://github.com/cla-assistant/github-action/issues/9) for this project. If you like to contribute your work, you can still raise a pull request. It is okay that you have rewritten almost all of the code. I will test all the use-cases and If it works fine, then I can merge your contribution 😄. Your Pull Request will be beneficial to this project and I can add you as a member of this organization.

**Comment by mattgreg** [comment]
We may have solved this problem in the Ockam fork 

https://github.com/ockam-network/cla-action

We are doing some final checks on the workflow and we should push it to main in the next day or two.

If the community likes our changes and updates, we are happy to submit a PR to move our added functionality upstream.

**Comment by ibakshay** [comment]
@mattgreg sounds great to me!!. Please go ahead and create a Pull Request. This functionality is one of the key functionalities which is missing in this CLA Assistant GitHub Action. So, your contribution will be very beneficial to the open-source community.

**Comment by ibakshay** [comment]
@mattgreg, Could you please update on your progress ?. Have you already implemented this feature ?. If yes, It will be great If you can submit a Pull Request 😄 . If not, I will develop this feature next week or so..

**Comment by mattgreg** [comment]
@ibakshay We have this working in production at https://github.com/ockam-network

The `cla-action` repo has the code we can submit with a PR to this upstream project.
Note the `contributors` repo and [the JSON file with the `allow` list of contributors](https://github.com/ockam-network/contributors/blob/master/cla-signers.json)

If you'd like to test the UX in our project, I'd suggest submitting a PR to fix a typo on our `website`.
[Here's one](https://www.ockam.io/learn/guides/contributions/ockam_c_code_standard/#use-of-goto-continue-break-and-). You can remove the `:` in the H3 of the .md [source](https://github.com/ockam-network/website/blob/production/src/content/learn/guides/contributions/ockam_c_code_standard.md#use-of-goto-continue-break-and-). None of the other H3's end in a colon so this one is a style typo.

**Comment by mattgreg** [comment]
Hey @ibakshay we were going to submit a PR today, but I see that you've done some work on this issue.

Do you want me to do anything with our fork?
Here is the commit with this feature: 
https://github.com/ockam-network/cla-action/commit/da5674edd7bbce94a78d9921b58e523dadb4548e

**Comment by ibakshay** [comment]
Hi @mattgreg, I actually introduced similar `remote-repo-signaturestorage-support` last week. I haven't updated on the readme file yet as I am still developing other features and bug fixes. However, thank you very much for coming forward to contribute to this repository :).

**Comment by mattgreg** [comment]
is there any part of the functionality the way we wrote it that you want us to contribute? We'd like to be in sync with your upstream branch.

**Comment by mattgreg** [comment]
I don't see us updating our fork to the upstream branch in this project in the foreseeable future. The functionality we built in our fork serves our purpose for the time being. We are focusing on core product development right now and won't be revisiting CLA automation anytime soon.

It was disappointing that the work we put into our fork wasn't able to be merged with your upstream project, as we anticipated, when we started development on this functionality. We don't feel like our efforts to move this project forward will be valued or additive to the progression of functionality in CLA Assistant. It may be best for you to go-it-alone with the project until it becomes a stable GA project. At that point we will upgrade, if appropriate.

I'm sorry that our collaboration didn't work out. We were excited about adding value and participating here.
